### PR TITLE
PCA9685 v2

### PR DESCRIPTION
**Pull request by ZioFabry**
## Description:

This version of the driver implement these new features:

- Support for multiple PCA9685 boards
- Support for a "simple" time based stepper/servo motor

**Related issue (if applicable):** fixes #<Tasmota issue number goes here>

## Checklist:
  - [x] The pull request is done against the latest development branch
  - [x] Only relevant files were touched
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [ ] The code change is tested and works with Tasmota core ESP8266 V.2.7.4.9
  - [x] The code change is tested and works with Tasmota core ESP32 V.2.0.9
  - [x] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

## Changelog Details

  - Code Refactoring
  - Support for multiple PCA9685 without breaking support for previuos driver version
    all command (PWM,INVERT,etc) now support the suffix from 0 to PCA9685_MAX_COUNT to address the board,
    without the suffix the command is relative to the board 0, e.g.
```
DRIVER15 PWM 0 350     // Board address 0 pin 0 value 350
DRIVER15 PWM0 2 300    // Board address 0 pin 2 value 300
DRIVER15 PWM1 3 235    // Board address 1 pin 3 value 235
DRIVER15 RESET2        // Reset Board 2
```
  - new command INTCLK to fine tuning the internal clock setting, unit: 0.1 MHz
    this is a not permanent setting!
```
DRIVER15 INTCLK 270    // Set Board #0 to 27.0 MHz
DRIVER15 INTCLK1 250   // Set Board #1 to 25.0 Mhz (default value, use USE_PCA9685_INT_CLOCK to change it at compile time)
```
  - new command PWMTO to move all the required pin from the current pin position to a new one, all the move will be completed into the required time, so each motor will be stepped relative to timing
  - new command PWMSTOP to stop all running movement
## Updated Command List:
```
DRIVER15 STATUS                   // Will return a JSON string containing all the current settings / parameters for all board
DRIVER15 RESET[0-8]               // Reset to power-up settings - i.e. F=50hz and all pins in OFF state for a specific board
DRIVER15 INVERT[0-8],pin[,0-1]    // print or set the inversion bit on pin of the specific board
DRIVER15 INTCLK[0-8],clock        // where clock is the Interal Clock value in 1/10 MHz (default USE_PCA9685_INT_CLOCK = 250)
DRIVER15 PMWF[0-8],frequency      // where frequency is the PWM frequency from 24 to 1526 in Hz
DRIVER15 PWM[0-8],pin,pwmvalue    // where pin=LED, pin 0 through 15 and pwmvalue is the pulse width between 0 and 4096
DRIVER15 PWM[0-8],pin,ON          // Fully turn a specific board/pin/LED ON
DRIVER15 PWM[0-8],pin,OFF         // Fully turn a specific board/pin/LED OFF
DRIVER15 PWMSTOP[0-8]             // stop all the moment on the relative board
DRIVER15 PWMTO[0-8] seconds,pin,value[,pin,value[,pin,value...]]  // Move all the specified pin to a new location in the specified time
```      
## PWMTO Example
```
PWMTO 4 0 327 1 550 2 187 3 200
```
this move the PIN0 of the Board 0 from the current position to 327
    the PIN1 of the Board 0 from the current position to 550
    the PIN2 of the Board 0 from the current position to 187
    the PIN3 of the Board 0 from the current position to 200
          
all the movements will be completed in 4 seconds, every PIN will be stepped relative to that @ 50ms step.
```
PWMTO1 4 0 327 1 550 2 187 3 200
```
same logic on the board 1

## TODO
- Permanent Settings
- Better Parameter Sanity check
- PWMRANGE command to set min/max PWM Value to be implemented

## Nice to have
- integrate the new PCA9685 driver into the Shutter code

### CODE REVIEW

File: tasmota/tasmota_xdrv_driver/xdrv_15_pca9685.ino [hunk 1]
```diff
@@ -25,165 +75,467 @@
...
+
+#include <Ticker.h>
+
+Ticker TickerPCA9685;
+
+void PCA9685_SetName(uint8_t pca)
+{
+  if (pca9685.count > 1)
+  {
+    pca9685.name[7] = '_';
```
Comment (arendst) [inline tasmota/tasmota_xdrv_driver/xdrv_15_pca9685.ino:135]: For consistency pls use function ``IndexSeparator()`` instead of '_'

**Comment by arendst** [comment]
Thx.

I'll add it as optional define USE_PCA9685_V2

**Comment by dr-apple** [comment]
It would be a dream if fading would be built in...

**Comment by ZioFabry** [comment]
You can use the PWMTO to fade the value.
The PWMTO start from the current pin value a move to the final value in the desidered window time... so the driver calculate what is the increment step to add/subtract from the current pwm value for each 50ms.

However is simple to implement a new command like:

PWMFADE[0-8] from,to,step,pin[,pin[,pin,...]] 

to fade all desidred pins 'from' -> 'to' adding the step value every 50ms... but probably is not needed.

**Comment by dr-apple** [comment]
Hello thank you very much, but the following commands not working:
15:49:32.342 CMD: DRIVER15 PWMFADE0 4096 0 3
15:49:32.347 RSL: RESULT = {"Command":"Error"}
15:49:39.872 CMD: DRIVER15 PWMFADE0 4096 0 
15:49:39.877 RSL: RESULT = {"Command":"Error"}

15:50:15.575 CMD: DRIVER15 PMWF0 120 
15:50:15.580 RSL: RESULT = {"Command":"Error"}

15:51:47.094 CMD: DRIVER15 PWMTO0 5 3 4096
15:51:47.100 RSL: RESULT = {"PCA9685":{"PWMTO":"OK"}}

here the problem is that it happens in the 5 seconds but rather in the direction of 15-20 seconds. And when it reaches the mx, the LED flashes briefly

**Comment by dr-apple** [comment]
ok because of the frequency, i found the error:
Command Upadte: DRIVER15 **PMWF**[0-8],frequency 

your write "PMWF" but you mean "PWMF" ;-)

**Comment by ZioFabry** [comment]
> Hello thank you very much, but the following commands not working: 15:49:32.342 CMD: DRIVER15 PWMFADE0 4096 0 3 15:49:32.347 RSL: RESULT = {"Command":"Error"} 15:49:39.872 CMD: DRIVER15 PWMFADE0 4096 0 15:49:39.877 RSL: RESULT = {"Command":"Error"}
> 
> 15:50:15.575 CMD: DRIVER15 PMWF0 120 15:50:15.580 RSL: RESULT = {"Command":"Error"}
> 
> 15:51:47.094 CMD: DRIVER15 PWMTO0 5 3 4096 15:51:47.100 RSL: RESULT = {"PCA9685":{"PWMTO":"OK"}}
> 
> here the problem is that it happens in the 5 seconds but rather in the direction of 15-20 seconds. And when it reaches the mx, the LED flashes briefly

The PWMFADE isn't implemented yet... I mean say that isn't complicated to implement it... If I found the time I will do it in the next days...

**Comment by ZioFabry** [comment]
> ok because of the frequency, i found the error: Command Upadte: DRIVER15 **PMWF**[0-8],frequency
> 
> your write "PMWF" but you mean "PWMF" ;-)

thanks... I will fix the comments.

**Comment by dr-apple** [comment]
> > Hello thank you very much, but the following commands not working: 15:49:32.342 CMD: DRIVER15 PWMFADE0 4096 0 3 15:49:32.347 RSL: RESULT = {"Command":"Error"} 15:49:39.872 CMD: DRIVER15 PWMFADE0 4096 0 15:49:39.877 RSL: RESULT = {"Command":"Error"}
> > 15:50:15.575 CMD: DRIVER15 PMWF0 120 15:50:15.580 RSL: RESULT = {"Command":"Error"}
> > 15:51:47.094 CMD: DRIVER15 PWMTO0 5 3 4096 15:51:47.100 RSL: RESULT = {"PCA9685":{"PWMTO":"OK"}}
> > here the problem is that it happens in the 5 seconds but rather in the direction of 15-20 seconds. And when it reaches the mx, the LED flashes briefly
> 
> The PWMFADE isn't implemented yet... I mean say that isn't complicated to implement it... If I found the time I will do it in the next days...

thank you very much, make it so: a config parameter to set the fadingtime for all or for 1 led , so we shelly at make

**Pull request closed without merging**